### PR TITLE
APIv3 blueprint: Add revisions endpoint

### PR DESCRIPTION
**Pull request by oliverguenther**
This commit provides the specification of a `revisions` APIv3 endpoint.
It is used to look up references from work packages made by revision
commit messages in the frontend.

Relevant work package: https://community.openproject.org/work_packages/19342

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-apiv3-revision-specification.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-apiv3-revision-specification.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: doc/apiv3-documentation.apib [hunk 1]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+
+# Group Revisions
+
+Revisions are sets of updates to files in the context of repositories linked in OpenProject.
+
+## Linked Properties
+|  Link            | Description                                         | Type          | Constraints | Supported operations |
+|:----------------:| --------------------------------------------------- | ------------- | ----------- | -------------------- |
+| self             | This revision                                       | Revision      | not null    | READ                 |
+| project          | The project to which the revision belongs           | Project       | not null    | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: In the long term this should rather be a repository, shouldn't it?

Or do you think that we can entirely leave out the concept of repositores from the APIv3 and replace it by projects? (serious question)

I am not quite sure which things a repository adds to a project (aside from revisions ^^).

My short-term question would be: Do we need a link to the project now, if we know that we will later _replace_ it with a different link?

File: doc/apiv3-documentation.apib [hunk 2]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+| committer        | The user that committed this revision, if defined   | User          |             | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: My first thought for this field name: Why is it not called `author`? Because that is the general name we use in the API for the person that creates something.

My second thought was: Or will it use [git semantics of committer vs author](http://stackoverflow.com/questions/18750808/difference-between-author-and-committer-in-git)? But if that was true more questions would arise, I therefore assume you are talking about the git author here.

**tl; dr:** I think this property should be called `author`

File: doc/apiv3-documentation.apib (hunk 1 above)
Comment (oliverguenther) [inline doc/apiv3-documentation.apib]: Assuming we're not implementing multiple repositories per project, the association `changeset - repository` and `changeset - project` is symmetric. If we decide to that a repository endpoint is required, we can link it **additionally** here.

Still, we can think about whether a repository endpoint would make sense. It could provide the following features:
- Branches
- Tags
- File access ? 
- Diff ?

File: doc/apiv3-documentation.apib (hunk 2 above)
Comment (oliverguenther) [inline doc/apiv3-documentation.apib]: I'm not even sure the repository backend would be consistent with the differing role of `author` and  `committer`, thus author is fine for me.

File: doc/apiv3-documentation.apib [hunk 3]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+
+
+## Local Properties
+| Property       | Description                                             | Type        | Constraints | Supported operations |
+|:--------------:| --------------------------------------------------------| ----------- | ----------- | -------------------- |
+| id             | Revisions's id                                          | Integer     | x > 0       | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: You already include the `identifier` for the SCM id, but in this description you _might_ want to clarify that this is the identifier assigned by OpenProject.

File: doc/apiv3-documentation.apib [hunk 4]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+| identifier     | The SCM identifier of the revision (e.g. SHA hash)      | String      | not null    | READ                 |
+| committerName  | The name of the author that committed this revision     | String      | not null    | READ                 |
+| message        | The commit message of the revision                      | Formattable | not null    | READ                 |
+| committedAt    | The time this revision was committed to the repository  | DateTime    | not null    | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: I'd again vote for `createdAt`, as this is APIv3-speak we are using in all other resources too.

I would only choose other words if you wanted to differentiate between (in git semantics) [commit date and the author date](http://stackoverflow.com/questions/11856983/why-git-authordate-is-different-from-commitdate).

In this case I'd probably use `createdAt` (because this is the APIv3 way to say it) and `commitedAt` (for the commit date in git semantics).

File: doc/apiv3-documentation.apib [hunk 5]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+| committer        | The user that committed this revision, if defined   | User          |             | READ                 |
+
+
+## Local Properties
+| Property       | Description                                             | Type        | Constraints | Supported operations |
+|:--------------:| --------------------------------------------------------| ----------- | ----------- | -------------------- |
+| id             | Revisions's id                                          | Integer     | x > 0       | READ                 |
+| identifier     | The SCM identifier of the revision (e.g. SHA hash)      | String      | not null    | READ                 |
+| committerName  | The name of the author that committed this revision     | String      | not null    | READ                 |
+| message        | The commit message of the revision                      | Formattable | not null    | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: :+1: for `Formattable`

File: doc/apiv3-documentation.apib [hunk 6]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+| project          | The project to which the revision belongs           | Project       | not null    | READ                 |
+| committer        | The user that committed this revision, if defined   | User          |             | READ                 |
+
+
+## Local Properties
+| Property       | Description                                             | Type        | Constraints | Supported operations |
+|:--------------:| --------------------------------------------------------| ----------- | ----------- | -------------------- |
+| id             | Revisions's id                                          | Integer     | x > 0       | READ                 |
+| identifier     | The SCM identifier of the revision (e.g. SHA hash)      | String      | not null    | READ                 |
+| committerName  | The name of the author that committed this revision     | String      | not null    | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: I first thought that this would duplicate the `committer`, but I see that we will probably not be able to match every SCM user to an OpenProject user.

Should we clarify that this is the SCM internal name of the user? (i.e. it can differ from the name the user has in OpenProject)

File: doc/apiv3-documentation.apib [hunk 7]
```diff
@@ -3155,6 +3233,7 @@ Note that due to sharing this might be more than the versions *defined* by that
 | category           | The category of the work package                                                                                                                      | Category     |              | READ / WRITE          |                                           |
 | priority           | The priority of the work package                                                                                                                      | Priority     | not null     | READ / WRITE          |                                           |
 | project            | The project to which the work package belongs                                                                                                         | Project      | not null     | READ                  |                                           |
+| revisions          | Revisions that are referencing the work package                                                                                                       | Revision     |              | READ                  | **Permission** view changesets            |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: DId you forget to specify the resource corresponding to this link? (e.g. `/api/v3/work_packages/42/revisions`)

Or do you plan to explicitly enumerate them inline? (I'd personally oppose doing that)

In any case you _should_ add an example to the WP resource below ^^

Comment (oliverguenther) [inline doc/apiv3-documentation.apib]: Ah, no - I was so busy with the endpoint that I forgot to fully describe it here.

File: doc/apiv3-documentation.apib (hunk 6 above)
Comment (oliverguenther) [inline doc/apiv3-documentation.apib]: We are not actively matching SCM users to OP users, but instead someone can actively _set_ the committer <-> user mapping in the repository settings. But I agree that this could use some more explicit description

File: doc/apiv3-documentation.apib (hunk 1 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: You are right... we can link _both_ in the future, there is no need to be stingy using links and project <-> revision is valid as long as we do not want to share repositories among projects ^^

:+1:

File: doc/apiv3-documentation.apib (hunk 6 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: :(

I thought we would be more awesome here ^^

File: doc/apiv3-documentation.apib [hunk 8]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+
+
+## Local Properties
+| Property       | Description                                                                                                                                                 | Type        | Constraints | Supported operations |
+|:--------------:| ------------------------------------------------------------------------------------------------------------------------------------------------------------| ----------- | ----------- | -------------------- |
+| id             | Revisions's id, assigned by OpenProject                                                                                                                     | Integer     | x > 0       | READ                 |
+| identifier     | The SCM identifier of the revision (e.g. SHA hash)                                                                                                          | String      | not null    | READ                 |
+| committerName  | The name of the author that committed this revision. Note that this name is retrieved from the repository and does not identify a user in OpenProject.      | String      | not null    | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: `authorName` ;)

File: doc/apiv3-documentation.apib [hunk 9]
```diff
@@ -2139,6 +2139,84 @@ For more details and all possible responses see the general specification of [Fo
...
+# Group Revisions
+
+Revisions are sets of updates to files in the context of repositories linked in OpenProject.
+
+## Linked Properties
+|  Link            | Description                                                                                       | Type          | Constraints | Supported operations |
+|:----------------:| --------------------------------------------------------------------------------------------------| ------------- | ----------- | -------------------- |
+| self             | This revision                                                                                     | Revision      | not null    | READ                 |
+| project          | The project to which the revision belongs                                                         | Project       | not null    | READ                 |
+| author           | The user that added this revision, if the committerName was mapped to a user in OpenProject       | User          |             | READ                 |
```
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib]: `authorName`

Comment (oliverguenther) [inline doc/apiv3-documentation.apib]: I disagree with `authorname` here, as it links to the actual user.
